control/control_test.go

```diff
@@ -136,127 +136,133 @@ func getPort() int {
 	c := New(getTestConfig())
-	err := c.Start()
-	Convey("pluginControl starts successfully", t, func() {
+	Convey("starts pluginControl", t, func() {
+		err := c.Start()
+		So(err, ShouldBeNil)
 		So(c.Started, ShouldBeTrue)
 		So(err, ShouldBeNil)
 		So(c.Name(), ShouldResemble, "control")
+		Convey("sets monitor duration", func() {
+			c.SetMonitorOptions(MonitorDurationOption(time.Millisecond * 100))
+			So(c.pluginRunner.Monitor().duration, ShouldResemble, 100*time.Millisecond)
+		})
+		c.Stop()
 	})
-	c.SetMonitorOptions(MonitorDurationOption(time.Millisecond * 100))
-	Convey("sets monitor duration", t, func() {
-		So(c.pluginRunner.Monitor().duration, ShouldResemble, 100*time.Millisecond)
-	})
-	c.Stop()
 }
 
 func TestSwapPlugin(t *testing.T) {
+	// These tests only work if SNAP_PATH is known.
+	// It is the responsibility of the testing framework to
+	// build the plugins first into the build dir.
 	if fixtures.SnapPath != "" {
-		c := New(getTestConfig())
-		c.Start()
-		time.Sleep(100 * time.Millisecond)
-		lpe := newListenToPluginEvent()
-		c.eventManager.RegisterHandler("Control.PluginsSwapped", lpe)
+		Convey("Starting plugin control", t, func() {
+			c := New(getTestConfig())
+			err := c.Start()
+			So(err, ShouldBeNil)
+			time.Sleep(100 * time.Millisecond)
 
-		_, e := load(c, fixtures.PluginPathMock2)
-		Convey("Loading first plugin", t, func() {
-			Convey("Should not error", func() {
-				So(e, ShouldBeNil)
-			})
-		})
-		// Travis optimization: If for some reason we can not load
-		// the plugin three times, we will fail the test immediately
-		// as we wait on a channel to be closed before proceeding with
-		// additional tests. If the plugin never loads, the channel will not
-		// close and just hang the test indefinitely.
-		if e != nil {
-			t.FailNow()
-		}
-		<-lpe.done
-		Convey("First plugin in catalog", t, func() {
-			Convey("Should have name mock", func() {
-				So(c.PluginCatalog()[0].Name(), ShouldEqual, "mock")
-			})
-		})
-		mockRP, mErr := core.NewRequestedPlugin(fixtures.PluginPathMock1)
-		Convey("Loading a plugin should not error", t, func() {
-			So(mErr, ShouldBeNil)
-		})
-		err := c.SwapPlugins(mockRP, c.PluginCatalog()[0])
-		Convey("Swapping plugins", t, func() {
-			Convey("Should not error", func() {
-				So(err, ShouldBeNil)
+			lpe := newListenToPluginEvent()
+			c.eventManager.RegisterHandler("Control.PluginsSwapped", lpe)
+
+			_, e := load(c, fixtures.PluginPathMock2)
+			Convey("Loading first plugin", func() {
+				Convey("Should not error", func() {
+					So(e, ShouldBeNil)
+				})
 			})
-		})
-		if err != nil {
-			t.FailNow()
-		}
-		<-lpe.done
 
-		// Swap plugin that was loaded with a different version of the plugin
-		Convey("Swapping plugins", t, func() {
-			Convey("Should generate a swapped plugins event", func() {
-				Convey("So first plugin in catalog after swap should have name mock", func() {
+			if e != nil {
+				t.FailNow()
```

Comment:
Having `t.Fatal("err =", e)` might be helpful in quickly determining why this eventually failed.